### PR TITLE
tests: add tests for `go/vt/hook`

### PR DESCRIPTION
**Pull request by Maniktherana**
## Description
Bump coverage from 37.7% to 78.3% 

## Related Issue(s)
#14931

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/hook/hook_test.go [hunk 1]
```diff
@@ -38,3 +45,223 @@ func TestExecuteContext(t *testing.T) {
...
+		{
+			name:       "HookSuccess",
+			hookName:   "echo",
+			parameters: []string{"test"},
+		},
+		{
+			name:          "HookDoesNotExist",
+			hookName:      "nonexistent-hook",
+			parameters:    []string{"test"},
+			expectedError: "",
```
Comment (frouioui) [inline go/vt/hook/hook_test.go:81]: nit: no need to set the field
```suggestion
```

File: go/vt/hook/hook_test.go [hunk 2]
```diff
@@ -38,3 +45,223 @@ func TestExecuteContext(t *testing.T) {
...
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			h := NewHook(tc.hookName, tc.parameters)
+			err := h.ExecuteOptional()
+			if tc.expectedError == "" {
+				assert.NoError(t, err)
+			} else {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tc.expectedError)
```
Comment (frouioui) [inline go/vt/hook/hook_test.go:93]: you can directly use `assert.ErrorContains` instead of using two distinct functions here

**Review by frouioui — COMMENTED** [review]
So far looks good to me besides the comments I left. I see coverage is up to almost 80%. More uncovered lines left are error cases, do you think you can go ahead and test those lines as well?

**Comment by Maniktherana** [comment]
> More uncovered lines left are error cases, do you think you can go ahead and test those lines as well?

@frouioui I'd rather tackle those in another PR as it's been slightly tricky trying to invoke stuff like `HOOK_GENERIC_ERROR` and `HOOK_TIMEOUT_ERROR`

**Comment by frouioui** [comment]
> I'd rather tackle those in another PR as it's been slightly tricky trying to invoke stuff like `HOOK_GENERIC_ERROR` and `HOOK_TIMEOUT_ERROR`

Sounds good to me!

**Comment by frouioui** [comment]
> Also one of the CI tests was failing. Think there was data race error for the `unit_race` test. Are there any changes I need to make for that?

It seems like `TestExecuteAsWritePipe` is flaky or failing when ran with `-race`, you should investigate locally to see if you can reproduce the issue. You can find the entire stack trace there: https://github.com/vitessio/vitess/actions/runs/7625023330/job/20769168942?pr=15015#logs

File: go/vt/hook/hook_test.go [hunk 3]
```diff
@@ -2,9 +2,11 @@ package hook
 
 import (
 	"context"
+	"io"
```
Comment (frouioui) [inline go/vt/hook/hook_test.go:5]: Also it seems like the license header is missing in this file, can you add it please?

File: go/vt/hook/hook_test.go [hunk 4]
```diff
@@ -38,3 +45,222 @@ func TestExecuteContext(t *testing.T) {
...
+	}()
+
+	h := NewHook("echo", nil)
+	var writer strings.Builder
+	writerPtr := &writer
+	_, waitFunc, status, err := h.ExecuteAsWritePipe(writerPtr)
+	require.NoError(t, err)
+	defer writerPtr.Reset()
+
+	_, err = writer.Write([]byte("Hello, World!\n"))
```
Comment (frouioui) [inline go/vt/hook/hook_test.go:250]: Yeah it seems like these lines of code are introducing the race condition, `writer` is not being locked

File: go/vt/hook/hook_test.go (hunk 3 above)
Comment (Maniktherana) [reply go/vt/hook/hook_test.go:21]: done

File: go/vt/hook/hook_test.go (hunk 4 above)
Comment (Maniktherana) [reply go/vt/hook/hook_test.go:266]: okay, I'll bring in a patch

Comment (Maniktherana) [reply go/vt/hook/hook_test.go]: @frouioui  done

**Review by frouioui — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]
LGTM

**Pull request merged**